pkg/workloadlearning/cache.go

```diff
@@ -0,0 +1,129 @@
+type ReadTableCostCache struct {
+	TableCostMetrics map[int64]*ReadTableCostMetrics // key: tableID
+	Version          uint64
+}
+
+// WLCacheWorker the worker to cache all workload-related metrics
+// Now it is also used to save the cache data of table cost metrics.
+type WLCacheWorker struct {
+	sysSessionPool     util.SessionPool
+	readTableCostCache *ReadTableCostCache
+	sync.RWMutex
+}
+
+// NewWLCacheWorker Create a new workload learning cache worker to cache all workload-related metrics
+// from storage mysql.tidb_workload_values to memory
+func NewWLCacheWorker(pool util.SessionPool) *WLCacheWorker {
+	return &WLCacheWorker{pool, &ReadTableCostCache{}, sync.RWMutex{}}
+}
+
+// UpdateTableCostCache refreshes the cached workload learning metrics
+func (cw *WLCacheWorker) UpdateTableCostCache() error {
+	// Get latest metrics from storage
+	se, err := cw.sysSessionPool.Get()
+	if err != nil {
+		return err
+	}
+	defer cw.sysSessionPool.Put(se)
+
+	sctx := se.(sessionctx.Context)
+	exec := sctx.GetRestrictedSQLExecutor()
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnWorkloadLearning)
+
+	// Whether to update table cost metrics
+	// Get the latest latestVersionInStorage in the storage
+	sql := `SELECT version FROM mysql.tidb_workload_values
+            WHERE category = %? AND type = %?
+            ORDER BY version DESC LIMIT 1`
+	rows, _, err := exec.ExecRestrictedSQL(ctx, nil, sql, feedbackCategory, tableCostType)
+	if err != nil {
+		return err
+	}
+	// Case: no metrics belongs to this feedback category and type
+	if len(rows) != 1 {
+		return nil
+	}
+	// If the latest latestVersionInStorage is the same as the cached latestVersionInStorage, no need to update
+	latestVersionInStorage := rows[0].GetUint64(0)
+	if latestVersionInStorage <= cw.readTableCostCache.Version {
+		return nil
+	}
+
+	// Get the latest table cost of metrics
+	sql = `SELECT table_id, value FROM mysql.tidb_workload_values
+            WHERE category = %? AND type = %? AND version = %?`
+	rows, _, err = exec.ExecRestrictedSQL(ctx, nil, sql, feedbackCategory, tableCostType, latestVersionInStorage)
+	if err != nil {
+		return err
+	}
+
+	newMetrics := make(map[int64]*ReadTableCostMetrics)
+	for _, row := range rows {
+		tableID := row.GetInt64(0)
+		value := row.GetBytes(1)
+
+		metric := &ReadTableCostMetrics{}
+		if err := json.Unmarshal(value, metric); err != nil {
+			logutil.BgLogger().Warn("failed to unmarshal table cost metrics",
+				zap.Int64("table_id", tableID),
+				zap.Error(err))
+			continue
+		}
+		newMetrics[tableID] = metric
+	}
+
+	// Update cache atomically
+	cw.RWMutex.Lock()
+	cw.readTableCostCache.TableCostMetrics = newMetrics
+	cw.readTableCostCache.Version = latestVersionInStorage
+	cw.RWMutex.Unlock()
+
+	return nil
+}
+
+// GetTableCostMetrics returns the cached metrics for a given table ID
+func (cw *WLCacheWorker) GetTableCostMetrics(tableID int64) *ReadTableCostMetrics {
```

Comment:
I think the priority queue should process all values at once. However, it’s fine to keep the current one and add a new method to retrieve all values.

Reply:
If you need to process all values at once, just don't forget to fetch and release the RWLock ~